File: src/main/java/io/mersel/dss/verify/api/models/TimestampInfo.java
package io.mersel.dss.verify.api.models;

import com.fasterxml.jackson.annotation.JsonInclude;

import java.util.Date;
import java.util.List;

/**
 * Zaman damgası bilgisi modeli
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class TimestampInfo {
    private boolean valid;
    private Date timestampTime;
    private String timestampType;
    private CertificateInfo tsaCertificate;
    private String digestAlgorithm;
    private String messageImprint;
    private String serialNumber;
    private String tsaName;
    private List<String> validationErrors;

    /**
     * Bu zaman damgasının DSS validation pipeline'ında <em>tek bir kök neden</em>
     * (root cause) constraint'i. Bkz. {@link SignatureInfo#getRootCause()}
     * dokümantasyonu — aynı sözleşmeyi taşır: pipeline-side-effect satırları
     * (XCV-top roll-up + SAV/CV cascade) sessizce filtrelenir, yalnız gerçek
     * kök neden döner. Birden fazla kök neden varsa DSS gezme sırasına göre
     * ilki seçilir; defansif fallback ile hiç kök neden tespit edilemezse
     * ham FAIL listesinden ilk satır seçilir. {@code null} → bu zaman
     * damgasında FAIL constraint yok (geçerli).
     *
     * <p>Eksiksiz audit için (tüm root cause'lar + roll-up/cascade satırları)
     * {@link #getFailedConstraints() failedConstraints} alanı opt-in olarak
     * doldurulabilir.</p>
     */
    private FailedConstraint rootCause;

    /**
     * Bu zaman damgasının DSS validation pipeline'ındaki <em>tüm</em> BBB
     * FAIL constraint'leri, {@link FailureCategory kategorize edilmiş} halde —
     * opt-in alan. Yalnız {@code ?includeFailedConstraints=true} query
     * parameter'i ile istendiğinde doldurulur; default <code>null</code>
     * kalır ({@code @JsonInclude(NON_NULL)} ile JSON'a yazılmaz).
     *
     * <p>Bkz. {@link SignatureInfo#getFailedConstraints()} dokümantasyonu —
     * aynı sözleşme: {@link FailureCategory#ROOT_CAUSE} (pipeline'ın
     * gerçek başarısızlık sebepleri), {@link FailureCategory#DERIVED}
     * (XCV-top summary roll-up), {@link FailureCategory#CASCADE}
     * (SAV/CV downstream yan ürün).</p>
     */
    private List<FailedConstraint> failedConstraints;

    // Getters and Setters
    public boolean isValid() {
        return valid;
    }

    public void setValid(boolean valid) {
        this.valid = valid;
    }

    public Date getTimestampTime() {
        return timestampTime;
    }

    public void setTimestampTime(Date timestampTime) {
        this.timestampTime = timestampTime;
    }

    public String getTimestampType() {
        return timestampType;
    }

    public void setTimestampType(String timestampType) {
        this.timestampType = timestampType;
    }

    public CertificateInfo getTsaCertificate() {
        return tsaCertificate;
    }

    public void setTsaCertificate(CertificateInfo tsaCertificate) {
        this.tsaCertificate = tsaCertificate;
    }

    public String getDigestAlgorithm() {
        return digestAlgorithm;
    }

    public void setDigestAlgorithm(String digestAlgorithm) {
        this.digestAlgorithm = digestAlgorithm;
    }

    public String getMessageImprint() {
        return messageImprint;
    }

    public void setMessageImprint(String messageImprint) {
        this.messageImprint = messageImprint;
    }

    public String getSerialNumber() {
        return serialNumber;
    }

    public void setSerialNumber(String serialNumber) {
        this.serialNumber = serialNumber;
    }

    public String getTsaName() {
        return tsaName;
    }

    public void setTsaName(String tsaName) {
        this.tsaName = tsaName;
    }

    public List<String> getValidationErrors() {
        return validationErrors;
    }

    public void setValidationErrors(List<String> validationErrors) {
        this.validationErrors = validationErrors;
    }

    public FailedConstraint getRootCause() {
        return rootCause;
    }

    public void setRootCause(FailedConstraint rootCause) {
        this.rootCause = rootCause;
    }

    public List<FailedConstraint> getFailedConstraints() {
        return failedConstraints;
    }

    public void setFailedConstraints(List<FailedConstraint> failedConstraints) {
        this.failedConstraints = failedConstraints;
    }
}

File: clients/dotnet-client/Models/TimestampInfo.cs
using System.Text.Json.Serialization;

namespace MERSEL.Services.DssVerifier.Client.Models;

/// <summary>
/// Bir imzaya bağlı zaman damgası (signature timestamp / archive timestamp)
/// bilgisi. Detached timestamp doğrulama için bkz.
/// <see cref="TimestampVerificationResult"/>.
/// </summary>
public sealed class TimestampInfo
{
    /// <summary>Token kriptografik olarak geçerli mi ve TSA sertifikası doğrulandı mı?</summary>
    [JsonPropertyName("valid")]
    public bool Valid { get; set; }

    /// <summary>TSA tarafından üretilen zaman damgası anı (genTime).</summary>
    [JsonPropertyName("timestampTime")]
    public DateTimeOffset? TimestampTime { get; set; }

    /// <summary>Timestamp türü (örn. <c>SIGNATURE_TIMESTAMP</c>, <c>ARCHIVE_TIMESTAMP</c>).</summary>
    [JsonPropertyName("timestampType")]
    public string? TimestampType { get; set; }

    /// <summary>Token'ı düzenleyen TSA sertifikası.</summary>
    [JsonPropertyName("tsaCertificate")]
    public CertificateInfo? TsaCertificate { get; set; }

    /// <summary>Message imprint hash algoritması (örn. <c>SHA-256</c>).</summary>
    [JsonPropertyName("digestAlgorithm")]
    public string? DigestAlgorithm { get; set; }

    /// <summary>Imzalı veriye ait hesaplanan message imprint (hex).</summary>
    [JsonPropertyName("messageImprint")]
    public string? MessageImprint { get; set; }

    /// <summary>TSA tarafından atanan seri numarası.</summary>
    [JsonPropertyName("serialNumber")]
    public string? SerialNumber { get; set; }

    /// <summary>TSA'nın insan-okur kimliği (Subject DN'inin CN veya tam adı).</summary>
    [JsonPropertyName("tsaName")]
    public string? TsaName { get; set; }

    /// <summary>Token doğrulanırken üretilen hata kayıtları.</summary>
    [JsonPropertyName("validationErrors")]
    public List<string>? ValidationErrors { get; set; }

    /// <summary>
    /// Bu zaman damgasının DSS validation pipeline'ında <em>tek bir kök neden</em>
    /// (root cause) BBB FAIL constraint'i.
    /// </summary>
    /// <remarks>
    /// <para>
    /// <see cref="SignatureInfo.RootCause"/> ile aynı sözleşmeyi taşır:
    /// pipeline-side-effect satırları (XCV-top roll-up + SAV/CV cascade) sessizce
    /// filtrelenir, yalnız gerçek kök neden döner. Birden fazla kök neden varsa
    /// DSS gezme sırasına göre ilki seçilir; defansif fallback ile hiç kök neden
    /// tespit edilemezse ham FAIL listesinden ilk satır seçilir.
    /// </para>
    /// <para>
    /// <c>null</c> → bu zaman damgasında FAIL constraint yok (geçerli).
    /// Eksiksiz audit için <see cref="FailedConstraints"/> alanı opt-in olarak
    /// doldurulabilir.
    /// </para>
    /// </remarks>
    /// <seealso cref="FailedConstraint"/>
    /// <seealso cref="FailedConstraints"/>
    [JsonPropertyName("rootCause")]
    public FailedConstraint? RootCause { get; set; }

    /// <summary>
    /// Bu zaman damgasının DSS validation pipeline'ındaki <em>tüm</em> BBB FAIL
    /// constraint'leri, <see cref="FailureCategory"/> ile kategorize edilmiş halde —
    /// opt-in alan.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Yalnız istekte <see cref="VerifySignatureRequest.IncludeFailedConstraints"/> =
    /// <c>true</c> verildiğinde doldurulur; default <c>null</c> kalır.
    /// Bkz. <see cref="SignatureInfo.FailedConstraints"/> — aynı sözleşmeyi taşır
    /// (ROOT_CAUSE + DERIVED + CASCADE kategorileri).
    /// </para>
    /// </remarks>
    /// <seealso cref="FailureCategory"/>
    /// <seealso cref="FailedConstraint"/>
    /// <seealso cref="VerifySignatureRequest.IncludeFailedConstraints"/>
    [JsonPropertyName("failedConstraints")]
    public List<FailedConstraint>? FailedConstraints { get; set; }
}
